File: subtitle/src/main/java/com/teixeira/subtitles/subtitle/utils/TimeUtils.java
package com.teixeira.subtitles.subtitle.utils;

/**
 * Contains functions for manipulating milliseconds and time strings.
 *
 * @author Felipe Teixeira
 */
public class TimeUtils {

  /**
   * Converts the given milliseconds to a String in the format hh:mm:ss,SSS.
   *
   * @param milliseconds Milliseconds to convert to String.
   * @return The time String in the format hh:mm:ss,SSS.
   */
  public static String getFormattedTime(long milliseconds) {
    long hours = milliseconds / 3600000;
    long minutes = (milliseconds % 3600000) / 60000;
    long seconds = (milliseconds % 60000) / 1000;
    long millis = milliseconds % 1000;

    return String.format("%02d:%02d:%02d,%03d", hours, minutes, seconds, millis);
  }

  /**
   * Converts the given hh:mm:ss,SSS format time string to milliseconds.
   *
   * @param time The time to be converted to milliseconds.
   * @return The milliseconds obtained from the time String.
   */
  public static long getMilliseconds(String time) {
    String[] timeParts = time.split(":");
    if (!isValidTime(timeParts)) {
      throw new IllegalArgumentException("Time format must be hh:mm:ss,SSS");
    }

    long hours = Long.parseLong(timeParts[0]);
    long minutes = Long.parseLong(timeParts[1]);

    String secondsAndMillis = timeParts[2];
    String[] secParts = secondsAndMillis.split(",");

    long seconds = Long.parseLong(secParts[0]);
    long milliseconds = Long.parseLong(secParts[1]);

    long totalMillis = (hours * 3600000) + (minutes * 60000) + (seconds * 1000) + milliseconds;

    return totalMillis;
  }

  /**
   * Checks if the time given in the format hh:mm:ss,SSS is valid.
   *
   * @param timeParts The parts of time divided by ":".
   * @return If the time provided is valid.
   */
  public static boolean isValidTime(String[] timeParts) {
    if (timeParts == null || timeParts.length != 3) {
      return false;
    }

    String secondsAndMillis = timeParts[2];
    String[] secParts = secondsAndMillis.split(",");
    if (secParts == null || secParts.length != 2) {
      return false;
    }

    try {
      int hours = Integer.parseInt(timeParts[0]);
      int minutes = Integer.parseInt(timeParts[1]);
      int seconds = Integer.parseInt(secParts[0]);
      int millis = Integer.parseInt(secParts[1]);

      if (hours < 0
          || hours > 99
          || minutes < 0
          || minutes > 59
          || seconds < 0
          || seconds > 59
          || millis < 0
          || millis > 999) {
        return false;
      }
    } catch (NumberFormatException e) {
      return false;
    }

    return true;
  }

  private TimeUtils() {}
}

File: app/src/main/java/com/teixeira0x/subtypo/utils/TimeUtils.kt
/*
 * This file is part of SubTypo.
 *
 * SubTypo is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * SubTypo is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with SubTypo.
 * If not, see <https://www.gnu.org/licenses/>.
 */

package com.teixeira0x.subtypo.utils

/**
 * Functions for dealing with times.
 *
 * @author Felipe Teixeira
 */
object TimeUtils {
  const val TIME_UNIT_HOUR = 3600000
  const val TIME_UNIT_MINUTE = 60000
  const val TIME_UNIT_SECOND = 1000

  /**
   * Converts the given milliseconds to a String in the format hh:mm:ss,SSS.
   *
   * @return The time String in the format hh:mm:ss,SSS.
   */
  fun Long.getFormattedTime(): String {
    val hours = this / TIME_UNIT_HOUR
    val minutes = (this % TIME_UNIT_HOUR) / TIME_UNIT_MINUTE
    val seconds = (this % TIME_UNIT_MINUTE) / TIME_UNIT_SECOND
    val millis = this % TIME_UNIT_SECOND

    return String.format("%02d:%02d:%02d,%03d", hours, minutes, seconds, millis)
  }

  /**
   * Converts the given hh:mm:ss,SSS format time string to milliseconds.
   *
   * @return The milliseconds obtained from the time String.
   */
  fun String.getMilliseconds(): Long {
    require(isValidTime(this)) { "Time format must be `hh:mm:ss,SSS`" }

    val timeParts = this.split(":")
    val hours = timeParts[0].toLong()
    val minutes = timeParts[1].toLong()

    val secondsAndMillis = timeParts[2]
    val secParts = secondsAndMillis.split(",")

    val seconds = secParts[0].toLong()
    val millis = secParts[1].toLong()

    val totalMillis =
      (hours * TIME_UNIT_HOUR) +
        (minutes * TIME_UNIT_MINUTE) +
        (seconds * TIME_UNIT_SECOND) +
        millis

    return totalMillis
  }

  /**
   * Checks if the time given in the format hh:mm:ss,SSS is valid.
   *
   * @param time The time to check.
   * @return If the time provided is valid.
   */
  fun isValidTime(time: String): Boolean {
    val timeParts = time.split(":")
    return runCatching {
        require(timeParts.size == 3) { "Time format must be `hh:mm:ss,SSS`" }

        val hours = timeParts[0]
        val minutes = timeParts[1]
        val secondsAndMillis = timeParts[2]
        val secParts = secondsAndMillis.split(",")
        require(secParts.size == 2) { "Seconds time format must be `ss,SSS`" }
        val seconds = secParts[0]
        val millis = secParts[1]

        val isValid =
          isInRange(hours, 0, 99, 2) &&
            isInRange(minutes, 0, 59, 2) &&
            isInRange(seconds, 0, 59, 2) &&
            isInRange(millis, 0, 999, 3)

        isValid
      }
      .getOrDefault(false)
  }

  /**
   * Checks if the given String has the expected length and if its numeric value
   * is within the specified bounds.
   *
   * @param value The String to be checked.
   * @param min The minimum numeric value (inclusive).
   * @param max The maximum numeric value (inclusive).
   * @param expectedLength The expected length of the String.
   * @return `true` if the numeric value of the String is within the bounds and
   *   its length matches the expected length, `false` otherwise.
   */
  fun isInRange(
    value: String,
    min: Int,
    max: Int,
    expectedLength: Int,
  ): Boolean {
    return value.length == expectedLength &&
      value.toIntOrNull()?.let { it in min..max } ?: false
  }
}
